### ISSUE TITLE
Dates not parsed correctly on Mac saved excel

### ISSUE THREAD

**Issue by coder36** [catamphetamine/read-excel-file#3]
Dates do not get parsed correctly if the use 1904 date system options is ticked, which it is by default on Macs.

![image](https://user-images.githubusercontent.com/561451/41086917-4768c44a-6a33-11e8-8104-a27e28ad716e.png)

https://support.microsoft.com/en-gb/help/214330/differences-between-the-1900-and-the-1904-date-system-in-excel

Awesome work BTW 👍Simplest .xslx parser I've used yet.

**Comment by catamphetamine** [comment]
can you post an XSLX file illustrating the issue?

**Comment by catamphetamine** [comment]
And the desired outcome, so that it could be put in a unit test.

**Comment by coder36** [comment]
The following spreadsheet has a column of dates set to 05/05/2018:
[example.xlsx](https://github.com/catamphetamine/read-excel-file/files/2080279/example.xlsx)

![image](https://user-images.githubusercontent.com/561451/41098201-77853338-6a52-11e8-8039-ad31571dfa00.png)


```
#!/usr/bin/env node
const readXlsxFile = require('read-excel-file/node');
const argv = require('yargs').usage('Usage: $0 --xlsx').demandOption(['xlsx']).argv;

const {xlsx} = argv

const schema = {
  'Date': {prop: 'date', type: Date },
}

async function run() {
  const {rows, errors} = await readXlsxFile(xlsx, { schema })
  console.log(rows)
}

run();
```
If I parse it using the above code, it outputs:

![image](https://user-images.githubusercontent.com/561451/41098126-40900222-6a52-11e8-851e-08ea10201c47.png)

It should be:

`date: 2018-05-05T12:00:00.000Z`

**Comment by catamphetamine** [comment]
Fixed in `read-excel-file@1.3.4`.

**Comment by catamphetamine** [comment]
Update:

Unpublished `read-excel-file@1.3.4` due to a bug.
Fixed in `read-excel-file@1.3.5`.

**Comment by catamphetamine** [comment]
Edit:

Unpublished `read-excel-file@1.3.5`.
Published `read-excel-file@1.3.6` (refactoring).

**Comment by catamphetamine** [comment]
By the way, I checked out XLSX spec a bit and seems that the date value could theoretically be parsed correctly by itself, without a schema, during the "read XLSX into an array of raw values" phase, because XLSX theoretically has some definition of dates.

For example, your XSLX file is:

```xml
<worksheet
	xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"
	xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" mc:Ignorable="x14ac"
	xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac">

	<sheetData>
		<row r="1">
			<c r="A1" t="s">
				<v>0</v>
			</c>
		</row>
		<row r="2">
			<c r="A2">
				<v>41763</v>
			</c>
		</row>
		<row r="3">
			<c r="A3">
				<f>A2</f>
				<v>41763</v>
			</c>
		</row>
		<row r="4">
			<c r="A4">
				<f t="shared" ref="A4:A6" si="0">A3</f>
				<v>41763</v>
			</c>
		</row>
		<row r="5">
			<c r="A5">
				<f t="shared" si="0" />
				<v>41763</v>
			</c>
		</row>
		<row r="6">
			<c r="A6" s="4">
				<f t="shared" si="0" />
				<v>41763</v>
			</c>
		</row>
	</sheetData>
</worksheet>
```

So, from the XML it can be seen that:

* The first row is a `s`hared `s`tring with index `0` (which is `"Date"`).
* The second row is just a `v`alue of `41763`.
* The third row is a `v`alue of `41763` also having `f`ormula same as for the second row (which is weird because the second row has no `f`ormula).
* The fourth row is a `v`alue of `41763` also having `f`ormula of `t`ype `"shared"` (strings) with index `0` (which is `"Date"`, so I guess they have a built-in "Date" formula or something) and also with `ref` (Range of Cells) equal to `A4:A6` which means that it "spans A4 to A6".

So, looks like XLSX is a messed up format, and parsing all those edge-cases (and how many of such cases not showcased in this sample file could exist) would be too much.
So I guess detecting and parsing Dates during the "read XLSX into an array of raw values" phase won't be implemented.
But `convertToJson` handles dates easily so I guess that will suffice.

**Comment by catamphetamine** [comment]
Ok, I think I found a way.

So, I will release a new (breaking) version of this library which will handle Booleans and Dates during the "parse raw data" stage (not requiring a schema).

Seems that a `<c/>`ell can have [`t="b"`](http://webapp.docx4java.org/OnlineDemo/ecma376/SpreadsheetML/ST_CellType.html) which would mean `boolean`.
So it will be parsed as `true`/`false` rather than `1`/`0` (and boolean parsing code will be removed from `convertToJson()`).

Second, `Date`s [seem to be parseable too](https://github.com/SheetJS/js-xlsx/blob/19620da30be2a7d7b9801938a0b9b1fd3c4c4b00/docbits/52_datatype.md).
XLSX doesn't define a dedicated `"d"`ate type but it has a hint of a date in "cell's style":

"cell's style" is the `s` attribute value (if present) and it's an index from `<cellXfs/>`:

```xml
 <cellXfs count="3">
  <xf numFmtId="0" fontId="0" fillId="0" borderId="0" xfId="0"/>
  <xf numFmtId="9" fontId="0" fillId="0" borderId="0" xfId="1" applyFont="1"/>
  <xf numFmtId="165" fontId="0" fillId="0" borderId="0" xfId="1" applyNumberFormat="1" applyFont="1"/>
 </cellXfs>
```

If [`numFmtId`](http://webapp.docx4java.org/OnlineDemo/ecma376/SpreadsheetML/numFmt.html) is `14-22` or `45-47` then it's a pre-defined date format so the `<v/>`alue is a date.

**Comment by catamphetamine** [comment]
So I played with it for a bit and found out that for your XLSX file it's:

```js
{ numberFormat: { id: '165', template: 'MM/DD/YY' } }
{ numberFormat: { id: '164', template: 'GENERAL' } }
{ numberFormat: { id: '164', template: 'GENERAL' } }
```

So it's neither of the pre-defined XLSX formats for some reason.
E.g. Google docs uses a custom format for Date too.

The library could examine the `template` like `if (template === 'MM/DD/YY')` but that would be English-only so instead I introduced an option: `dateFormat: 'MM/DD/YY'` — if it's passed then such dates are parsed during the "raw data" stage. In any case, dates already get parsed during `convertToJson()` so no changes there.

Released v `2.0.0`.
If you're only using `readXlsx()` with `schema` parameter then there's no difference.
The only difference is for using `readXlsx()` without `schema` parameter.
